File: typescript/BestTimeBuySellStock.ts
export class BestTimeBuySellStock {
  //console.log(maxProfit([7, 1, 5, 3, 6, 4]));

  maxProfit(prices: number[]): number {
    let minValue = Number.MAX_VALUE;
    let maxProfit = 0;

    for (let i = 0; i < prices.length; i++) {
      if (prices[i] < minValue) minValue = prices[i];
      else if (prices[i] - minValue > maxProfit) {
        maxProfit = prices[i] - minValue;
      }
    }

    return maxProfit;
  }
}

File: src/main/java/BestTimeBuySellStock.java
public class BestTimeBuySellStock {

    public static void main(String args[]) {
        int array[] = { 3, 3, 5, 0, 0, 3, 1, 4 };

        System.out.println(maxProfit(array));

    }

    public static int maxProfit(int[] prices) {
        int minValue = Integer.MAX_VALUE;
        int maxProfit = 0;

        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < minValue)
                minValue = prices[i];

            else if (prices[i] - minValue > maxProfit) {
                maxProfit = prices[i] - minValue;

            }

        }

        return maxProfit;

    }

}
